File: lib/introjs.dart
@JS()
library introjs;

import "dart:html";
import "dart:async";

import "package:js/js.dart";

/**
 * Barebones introJs wrapper class
 *
 * Ideally would be named _introJs but that's currently impossible due to https://github.com/dart-lang/sdk/issues/27537
 */
@JS("introJs")
class InternalIntroJs {
  external InternalIntroJs([String element]);

  external InternalIntroJs clone();
  external InternalIntroJs setOption(String option, var value);
  external InternalIntroJs setOptions(Options options);
  external InternalIntroJs start();
  external InternalIntroJs goToStep(int step);
  external InternalIntroJs nextStep();
  external InternalIntroJs previousStep();
  external InternalIntroJs exit();
  external InternalIntroJs refresh();
  external InternalIntroJs onbeforechange(
      void providedCallback(Element element));
  external InternalIntroJs onchange(void providedCallback(Element element));
  external InternalIntroJs onafterchange(
      void providedCallback(Element element));
  external InternalIntroJs oncomplete(void providedCallback());
  external InternalIntroJs onhintsadded(void providedCallback());
  external InternalIntroJs onhintclick(
      void providedCallback(Element hintElement, var item, int stepId));
  external InternalIntroJs onhintclose(void providedCallback(int stepId));
  external InternalIntroJs onexit(void providedCallback());
  external InternalIntroJs addHints();
  external InternalIntroJs hideHint(int stepId);
  external InternalIntroJs hideHints();
}

class HintAddedEvent {
  Element hintElement;
  Hint item;
  int stepId;

  HintAddedEvent(this.hintElement, this.item, this.stepId);
}

/**
 * Wraps [InternalIntroJs] and offers a Dart friendly interface
 */
class IntroJs {
  InternalIntroJs _jsObj;

  IntroJs([String element]) {
    if (element != null) {
      _jsObj = new InternalIntroJs(element);
    } else {
      _jsObj = new InternalIntroJs();
    }
  }

  void setOption(String option, var value) {
    _jsObj.setOption(option, value);
  }

  void setOptions(Options options) {
    _jsObj.setOptions(options);
  }

  void start() {
    _jsObj.start();
  }

  void goToStep(int step) {
    _jsObj.goToStep(step);
  }

  void nextStep() {
    _jsObj.nextStep();
  }

  void previousStep() {
    _jsObj.previousStep();
  }

  void exit() {
    _jsObj.exit();
  }

  void refresh() {
    _jsObj.refresh();
  }

  Stream<Element> get onBeforeChange {
    StreamController<Element> controller = new StreamController<Element>();
    _jsObj.onbeforechange(
        allowInterop((Element element) => controller.add(element)));
    return controller.stream;
  }

  Stream<Element> get onChange {
    StreamController<Element> controller = new StreamController<Element>();
    _jsObj.onchange(allowInterop((Element element) => controller.add(element)));
    return controller.stream;
  }

  Stream<Element> get onAfterChange {
    StreamController<Element> controller = new StreamController<Element>();
    _jsObj.onafterchange(
        allowInterop((Element element) => controller.add(element)));
    return controller.stream;
  }

  Future get onComplete {
    Completer completer = new Completer();
    _jsObj.oncomplete(allowInterop(() => completer.complete(null)));
    return completer.future;
  }

  Future get onHintsAdded {
    Completer completer = new Completer();
    _jsObj.onhintsadded(allowInterop(() => completer.complete(null)));
    return completer.future;
  }

  Stream<HintAddedEvent> get onHintClick {
    StreamController<HintAddedEvent> controller =
        new StreamController<HintAddedEvent>();
    _jsObj.onhintclick(allowInterop(
        (Element hintElement, var item, int stepId) => controller.add(
            new HintAddedEvent(
                hintElement,
                new Hint(
                    element: item.element,
                    hint: item.hint,
                    hintAnimation: item.hintAnimation,
                    targetElement: item.targetElement),
                stepId))));
    return controller.stream;
  }

  Stream<int> get onHintClose {
    StreamController<int> controller = new StreamController<int>();
    _jsObj.onhintclose(allowInterop((int stepId) => controller.add(stepId)));
    return controller.stream;
  }

  Future get onExit {
    Completer completer = new Completer();
    _jsObj.onexit(allowInterop(() => completer.complete(null)));
    return completer.future;
  }

  void addHints() {
    _jsObj.addHints();
  }

  void hideHint(int stepId) {
    _jsObj.hideHint(stepId);
  }

  void hideHints() {
    _jsObj.hideHints();
  }
}

@JS()
@anonymous
class Options {
  external String get nextLabel;
  external String get prevLabel;
  external String get skipLabel;
  external String get doneLabel;
  external bool get hidePrev;
  external bool get hideNext;
  external String get tooltipPosition;
  external String get tooltipClass;
  external String get highlightClass;
  external bool get exitOnEsc;
  external bool get exitOnOverlayClick;
  external bool get showStepNumbers;
  external bool get keyboardNavigation;
  external bool get showButtons;
  external bool get showBullets;
  external bool get showProgress;
  external bool get scrollToElement;
  external num get overlayOpacity;
  external num get scrollPadding;
  external List<String> get positionPrecedence;
  external bool get disableInteraction;
  external String get hintPosition;
  external String get hintButtonLabel;
  external bool get hintAnimation;
  external List<Hint> get hints;

  external factory Options(
      {String nextLabel,
      String prevLabel,
      String skipLabel,
      String doneLabel,
      bool hidePrev,
      bool hideNext,
      String tooltipPosition,
      String tooltipClass,
      String highlightClass,
      bool exitOnEsc,
      bool exitOnOverlayClick,
      bool showStepNumbers,
      bool keyboardNavigation,
      bool showButtons,
      bool showBullets,
      bool showProgress,
      bool scrollToElement,
      num overlayOpacity,
      num scrollPadding,
      List<String> positionPrecedence,
      bool disableInteraction,
      String hintPosition,
      String hintButtonLabel,
      bool hintAnimation,
      List<Hint> hints,
      List<Step> steps});
}

@JS()
@anonymous
class Hint {
  external dynamic get element;
  external String get hint;
  external String get hintPosition;
  external bool get hintAnimation;
  external String get tooltipClass;
  external String get position;
  external String get targetElement;

  external factory Hint(
      {var element,
      String hint,
      String hintPosition,
      bool hintAnimation,
      String tooltipClass,
      String position,
      String targetElement});
}

@JS()
@anonymous
class Step {
  external dynamic get element;
  external String get intro;
  external String get position;

  external factory Step({var element, String intro, String position});
}

File: src/IntroJs.ts
import mergeOptions from "./util/mergeOptions";
import exitIntro from "./core/exitIntro";
import refresh from "./core/refresh";
import introForElement from "./core/introForElement";

import {
  populateHints,
  hideHint,
  hideHints,
  showHint,
  showHints,
  removeHint,
  removeHints,
  showHintDialog,
} from "./core/hint";
import {
  currentStep,
  goToStep,
  goToStepNumber,
  nextStep,
  previousStep,
} from "./core/steps";
import { getDontShowAgain, setDontShowAgain } from "./core/dontShowAgain";

export interface IntroItem {
  element: HTMLElement;
  title?: string | null;
  intro?: string | null;
  step?: number;
  hint?: string;
  hintPosition?: string;
  hintAnimation?: boolean;
  tooltipClass?: string | null;
  highlightClass?: string | null;
  position?: string;
  scrollTo?: string;
  disableInteraction?: boolean;
}

export interface Options {
  isActive?: boolean;
  showButtons?: boolean;
  scrollPadding?: number;
  prevLabel?: string;
  showProgress?: boolean;
  hintButtonLabel?: string;
  positionPrecedence?: string[];
  progressBarAdditionalClass?: boolean;
  doneLabel?: string;
  hideNext?: boolean;
  hidePrev?: boolean;
  highlightClass?: string;
  showStepNumbers?: boolean;
  keyboardNavigation?: boolean;
  overlayOpacity?: number;
  disableInteraction?: boolean;
  exitOnEsc?: boolean;
  exitOnOverlayClick?: boolean;
  tooltipClass?: string;
  showBullets?: boolean;
  hintPosition?: string;
  scrollToElement?: boolean;
  scrollTo?: string;
  helperElementPadding?: number;
  skipLabel?: string;
  nextToDone?: boolean;
  buttonClass?: string;
  nextLabel?: string;
  tooltipPosition?: string;
  hintAnimation?: boolean;
  dontShowAgainCookie?: string;
  dontShowAgainCookieDays?: number;
  steps?: Array<Step>;
  hints?: Array<any>;
  group?: string;
  autoPosition?: boolean;
  dontShowAgain?: boolean;
  dontShowAgainLabel?: string;
  stepNumbersOfLabel?: string;
  hintShowButton?: boolean;
  hintAutoRefreshInterval?: number;
}

export interface Step {
  element?: HTMLElement | HTMLDivElement | HTMLParagraphElement | string | null;
  intro: string;
  position?:
    | "top"
    | "right"
    | "left"
    | "bottom"
    | "bottom-left-aligned"
    | "bottom-middle-aligned"
    | "bottom-right-aligned"
    | "auto";
}

export type IntroChangeCallback = (
  this: IntroJs,
  element?: HTMLElement
) => boolean;
export type IntroBeforeChangeCallback = (
  this: IntroJs,
  element?: any
) => boolean;
export type IntroAfterChangeCallback = (
  this: IntroJs,
  element?: HTMLElement
) => void;
export type IntroCompleteCallback = (
  this: IntroJs,
  currentStep: number | undefined,
  status: string
) => void;
export type IntroExitCallback = (this: IntroJs) => void;
export type IntroSkipCallback = (this: IntroJs) => void;
export type IntroStartCallback = (this: IntroJs, element?: HTMLElement) => void;
export type IntroBeforeExitCallback = (this: IntroJs) => boolean;
export type HintsAddedCallback = (this: IntroJs) => void;
export type HintsAutoRefreshFunction = (this: IntroJs) => void;
export type HintClickCallback = (
  hintElement?: HTMLElement,
  item?: IntroItem,
  stepId?: number
) => void;
export type HintCloseCallback = (
  this: IntroJs,
  step?: number | undefined | string
) => void;

export class IntroJs {
  _options: Options;
  _introItems: IntroItem[] = [];
  _currentStep: number | undefined = undefined;
  _direction: string | undefined = undefined;
  _currentStepNumber: number | undefined = undefined;
  _lastShowElementTimer: number | undefined = undefined;
  _targetElement: HTMLElement;

  _introBeforeChangeCallback: IntroBeforeChangeCallback | undefined;
  _introChangeCallback: IntroChangeCallback | undefined;
  _introAfterChangeCallback: IntroAfterChangeCallback | undefined;
  _introCompleteCallback: IntroCompleteCallback | undefined;
  _hintsAddedCallback: HintsAddedCallback | undefined;
  _hintClickCallback: HintClickCallback | undefined;
  _hintCloseCallback: HintCloseCallback | undefined;
  _introExitCallback: IntroExitCallback | undefined;
  _introProvidedCallback: IntroStartCallback | undefined;
  _introSkipCallback: IntroSkipCallback | undefined;
  _introStartCallback: IntroStartCallback | undefined;
  _introBeforeExitCallback: IntroBeforeExitCallback | undefined;
  _hintsAutoRefreshFunction: HintsAutoRefreshFunction | undefined;

  constructor(obj: HTMLElement) {
    this._targetElement = obj;
    this._introItems = [];

    this._options = {
      /* Is this tour instance active? Don't show the tour again if this flag is set to false */
      isActive: true,
      /* Next button label in tooltip box */
      nextLabel: "Next",
      /* Previous button label in tooltip box */
      prevLabel: "Back",
      /* Skip button label in tooltip box */
      skipLabel: "×",
      /* Done button label in tooltip box */
      doneLabel: "Done",
      /* Hide previous button in the first step? Otherwise, it will be disabled button. */
      hidePrev: false,
      /* Hide next button in the last step? Otherwise, it will be disabled button (note: this will also hide the "Done" button) */
      hideNext: false,
      /* Change the Next button to Done in the last step of the intro? otherwise, it will render a disabled button */
      nextToDone: true,
      /* Default tooltip box position */
      tooltipPosition: "bottom",
      /* Next CSS class for tooltip boxes */
      tooltipClass: "",
      /* Start intro for a group of elements */
      group: "",
      /* CSS class that is added to the helperLayer */
      highlightClass: "",
      /* Close introduction when pressing Escape button? */
      exitOnEsc: true,
      /* Close introduction when clicking on overlay layer? */
      exitOnOverlayClick: true,
      /* Show step numbers in introduction? */
      showStepNumbers: false,
      /* Let user use keyboard to navigate the tour? */
      keyboardNavigation: true,
      /* Show tour control buttons? */
      showButtons: true,
      /* Show tour bullets? */
      showBullets: true,
      /* Show tour progress? */
      showProgress: false,
      /* Scroll to highlighted element? */
      scrollToElement: true,
      /* "Don't show again" cookie name and expiry (in days) */
      dontShowAgainCookie: "introjs-dontShowAgain",
      dontShowAgainCookieDays: 365,
      /* Pagination "of" label */
      stepNumbersOfLabel: "of",
      /*
       * Should we scroll the tooltip or target element?
       *
       * Options are: 'element' or 'tooltip'
       */
      scrollTo: "element",
      /* Padding to add after scrolling when element is not in the viewport (in pixels) */
      scrollPadding: 30,
      /* Set the overlay opacity */
      overlayOpacity: 0.5,
      /* Precedence of positions, when auto is enabled */
      positionPrecedence: ["bottom", "top", "right", "left"],
      /* Disable an interaction with element? */
      disableInteraction: false,
      /* Set how much padding to be used around helper element */
      helperElementPadding: 10,
      /* Default hint position */
      hintPosition: "top-middle",
      /* Hint button label */
      hintButtonLabel: "Got it",
      /* Adding animation to hints? */
      hintAnimation: true,
      /* additional classes to put on the buttons */
      buttonClass: "introjs-button",
      /* additional classes to put on progress bar */
      progressBarAdditionalClass: false,
      /* To determine the tooltip position automatically based on the window.width/height */
      autoPosition: true,
      /* To display the "Don't show again" checkbox in the tour */
      dontShowAgain: false,
      dontShowAgainLabel: "Don't show this again",
      /* Display the "Got it" button? */
      hintShowButton: true,
      /* Hints auto-refresh interval in ms (set to -1 to disable) */
      hintAutoRefreshInterval: 10,
    };
  }
  isActive() {
    if (this._options.dontShowAgain && getDontShowAgain.call(this)) {
      return false;
    }

    return this._options.isActive;
  }

  clone() {
    return new IntroJs(this._targetElement);
  }
  setOption(option: string, value: any) {
    // @ts-ignore
    this._options[option] = value;
    return this;
  }
  setOptions(options: Options) {
    this._options = mergeOptions(this._options, options);
    return this;
  }
  start() {
    introForElement.call(this, this._targetElement);
    return this;
  }
  goToStep(step: number) {
    goToStep.call(this, step);
    return this;
  }
  addStep(options: Step) {
    if (!this._options.steps) {
      this._options.steps = [];
    }

    this._options.steps.push(options);

    return this;
  }
  addSteps(steps: Array<Step>) {
    if (!steps.length) return;

    for (let index = 0; index < steps.length; index++) {
      this.addStep(steps[index]);
    }

    return this;
  }
  goToStepNumber(step: number) {
    goToStepNumber.call(this, step);

    return this;
  }
  nextStep() {
    nextStep.call(this);
    return this;
  }
  previousStep() {
    previousStep.call(this);
    return this;
  }
  currentStep() {
    return currentStep.call(this);
  }
  exit(force: boolean = false) {
    exitIntro.call(this, this._targetElement, force);
    return this;
  }
  refresh(refreshSteps: boolean = false) {
    refresh.call(this, refreshSteps);
    return this;
  }
  setDontShowAgain(dontShowAgain: boolean) {
    setDontShowAgain.call(this, dontShowAgain);
    return this;
  }
  onbeforechange(providedCallback: IntroBeforeChangeCallback) {
    if (typeof providedCallback === "function") {
      this._introBeforeChangeCallback = providedCallback;
    } else {
      throw new Error(
        "Provided callback for onbeforechange was not a function"
      );
    }
    return this;
  }
  onchange(providedCallback: IntroChangeCallback) {
    if (typeof providedCallback === "function") {
      this._introChangeCallback = providedCallback;
    } else {
      throw new Error("Provided callback for onchange was not a function.");
    }
    return this;
  }
  onafterchange(providedCallback: IntroAfterChangeCallback) {
    if (typeof providedCallback === "function") {
      this._introAfterChangeCallback = providedCallback;
    } else {
      throw new Error("Provided callback for onafterchange was not a function");
    }
    return this;
  }
  oncomplete(providedCallback: IntroCompleteCallback) {
    if (typeof providedCallback === "function") {
      this._introCompleteCallback = providedCallback;
    } else {
      throw new Error("Provided callback for oncomplete was not a function.");
    }
    return this;
  }
  onhintsadded(providedCallback: HintsAddedCallback) {
    if (typeof providedCallback === "function") {
      this._hintsAddedCallback = providedCallback;
    } else {
      throw new Error("Provided callback for onhintsadded was not a function.");
    }
    return this;
  }
  onhintclick(providedCallback: HintClickCallback) {
    if (typeof providedCallback === "function") {
      this._hintClickCallback = providedCallback;
    } else {
      throw new Error("Provided callback for onhintclick was not a function.");
    }
    return this;
  }
  onhintclose(providedCallback: HintCloseCallback) {
    if (typeof providedCallback === "function") {
      this._hintCloseCallback = providedCallback;
    } else {
      throw new Error("Provided callback for onhintclose was not a function.");
    }
    return this;
  }
  onstart(providedCallback: IntroStartCallback) {
    if (typeof providedCallback === "function") {
      this._introStartCallback = providedCallback;
    } else {
      throw new Error("Provided callback for onstart was not a function.");
    }
    return this;
  }
  onexit(providedCallback: IntroExitCallback) {
    if (typeof providedCallback === "function") {
      this._introExitCallback = providedCallback;
    } else {
      throw new Error("Provided callback for onexit was not a function.");
    }
    return this;
  }
  onskip(providedCallback: IntroSkipCallback) {
    if (typeof providedCallback === "function") {
      this._introSkipCallback = providedCallback;
    } else {
      throw new Error("Provided callback for onskip was not a function.");
    }
    return this;
  }
  onbeforeexit(providedCallback: IntroBeforeExitCallback) {
    if (typeof providedCallback === "function") {
      this._introBeforeExitCallback = providedCallback;
    } else {
      throw new Error("Provided callback for onbeforeexit was not a function.");
    }
    return this;
  }
  addHints() {
    populateHints.call(this, this._targetElement);
    return this;
  }
  hideHint(stepId: number) {
    hideHint.call(this, stepId);
    return this;
  }
  hideHints() {
    hideHints.call(this);
    return this;
  }
  showHint(stepId: number) {
    showHint.call(this, stepId);
    return this;
  }
  showHints() {
    showHints.call(this);
    return this;
  }
  removeHints() {
    removeHints.call(this);
    return this;
  }
  removeHint(stepId: number) {
    removeHint.call(this, stepId);
    return this;
  }
  showHintDialog(stepId: number) {
    showHintDialog.call(this, stepId);
    return this;
  }
}
